tests/unit/clients/test_gs_client.py

```diff
@@ -0,0 +1,208 @@
+"""Test of the GS Client."""
+import io
+
+import pytest
+import mock
+
+from google.cloud import exceptions as google_exceptions
+
+from tentaclio.clients import gs_client, exceptions
+
+
+@mock.patch("tentaclio.clients.GSClient._connect")
+@pytest.mark.parametrize(
+    "url,hostname,path",
+    [
+        ("gs://bucket/prefix", "bucket", "prefix"),
+        ("gs://:@gs", None, ""),
+        ("gs://public_key:private_key@gs", None, ""),
+        ("gs://:@bucket", "bucket", ""),
+        ("gs://:@bucket/prefix", "bucket", "prefix"),
+    ],
+)
+def test_parsing_gs_url(m_connect, url, hostname, path):
+    """Test the parsing of the gs url."""
+    client = gs_client.GSClient(url)
+
+    assert client.key_name == path
+    assert client.bucket == hostname
+
+
+def patch_string_from_method(method: str) -> str:
+    """Get the string that indicates the method to be patched for a calling method."""
+    switch_dict = {
+        "get": "tentaclio.clients.GSClient._get",
+        "put": "tentaclio.clients.GSClient._put",
+        "remove": "tentaclio.clients.GSClient._remove"
+    }
+
+    if method in switch_dict:
+        return switch_dict[method]
+
+    raise Exception(f"No patch string for calling method: {method}")
+
+
+@mock.patch("tentaclio.clients.GSClient._connect")
+@pytest.mark.parametrize(
+    "method,url,bucket,key", [
+        ("get", "gs://:@gs", None, None),
+        ("get", "gs://:@gs", "bucket", None),
+        ("get", "gs://:@bucket", None, None),
+        ("put", "gs://:@gs", None, None),
+        ("put", "gs://:@gs", "bucket", None),
+        ("put", "gs://:@bucket", None, None),
+        ("remove", "gs://:@gs", None, None),
+        ("remove", "gs://:@gs", "bucket", None),
+        ("remove", "gs://:@bucket", None, None)
+    ],
+)
+def test_invalid_path(m_connect, method, url, bucket, key):
+    """Test a method with invalid paths."""
+    patch_string = patch_string_from_method(method)
+
+    with mock.patch(patch_string) as mocked_method:
+        with gs_client.GSClient(url) as client:
+            with pytest.raises(exceptions.GSError):
```

Comment:
FYI you can combine context managers into one line so you don't have to do 3 indents, and the mock can be safely done without context manager because it's only used as mock throughout the test:
```
mocked_method = mocker.patch(patch_string)
with gs_client.GSClient(url) as client, pytest.raises(exceptions.GSError):
    ...
```